src/mprisplayer.cpp

```diff
@@ -565,19 +568,24 @@ QStringList MprisPlayer::supportedMimeTypes() const
 void MprisPlayer::setServiceName(const QString &serviceName)
 {
     if (!priv->m_serviceName.isEmpty()) {
-        priv->m_connection.unregisterObject(QStringLiteral("/org/mpris/MediaPlayer2"));
-        priv->m_connection.unregisterService(priv->m_serviceName);
+        priv->m_connection->unregisterObject(QStringLiteral("/org/mpris/MediaPlayer2"));
+        priv->m_connection->unregisterService(priv->m_serviceName);
+        QDBusConnection::disconnectFromBus(priv->m_serviceName);
+        delete priv->m_connection;
+        priv->m_connection = nullptr;
     }
 
     if (!serviceName.isEmpty()) {
+        priv->m_connection = new QDBusConnection(QDBusConnection::connectToBus(dbusConnectionType(), serviceName));
+
         if (!serviceName.startsWith(QLatin1String("org.mpris.MediaPlayer2."))) {
             priv->m_serviceName = QStringLiteral("%1.%2.instance%3").arg(QLatin1String("org.mpris.MediaPlayer2")).arg(serviceName).arg(QCoreApplication::applicationPid());
```

Comment:
The instance id stuff is moved to the DeclarativeMprisPlayer in the second commit, so that service name pattern is only used by QML apps. The C++ API gives more control over the service name to avoid that problem.